pkg/operator/resource/resourceapply/monitoring.go

```diff
@@ -2,148 +2,168 @@ package resourceapply
-	if errors.IsNotFound(err) {
-		newObj, createErr := client.Resource(prometheusRuleGVR).Namespace(namespace).Create(ctx, required, metav1.CreateOptions{})
-		if createErr != nil {
-			recorder.Warningf("PrometheusRuleCreateFailed", "Failed to create PrometheusRule.monitoring.coreos.com/v1: %v", createErr)
-			return nil, true, createErr
-		}
-		recorder.Eventf("PrometheusRuleCreated", "Created PrometheusRule.monitoring.coreos.com/v1 because it was missing")
-		return newObj, true, nil
+	// Cast the metadata to the correct type.
+	var existingObjectMetaTyped, requiredObjectMetaTyped metav1.ObjectMeta
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(existingObjectMeta, &existingObjectMetaTyped)
+	if err != nil {
+		return nil, false, err
 	}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(requiredObjectMeta, &requiredObjectMetaTyped)
 	if err != nil {
 		return nil, false, err
 	}
 
-	existingCopy := existing.DeepCopy()
+	// Check if the metadata objects differ.
+	didMetadataModify := ptr.To(false)
+	resourcemerge.EnsureObjectMeta(didMetadataModify, &existingObjectMetaTyped, requiredObjectMetaTyped)
 
-	toUpdate, modified, err := ensureGenericSpec(required, existingCopy, noDefaulting, equality.Semantic)
+	// Deep-check the spec objects for equality, and update the cache in either case.
+	if defaultingFunc == nil {
+		defaultingFunc = noDefaulting
+	}
+	if equalityChecker == nil {
+		equalityChecker = equality.Semantic
+	}
+	existingCopy, didSpecModify, err := ensureGenericSpec(required, existingCopy, defaultingFunc, equalityChecker)
 	if err != nil {
 		return nil, false, err
 	}
-
-	if !modified {
-		return nil, false, nil
+	if !didSpecModify && !*didMetadataModify {
+		// Update cache even if certain fields are not modified, in order to maintain a consistent cache based on the
+		// resource hash. The resource hash depends on the entire metadata, not just the fields that were checked above,
+		cache.UpdateCachedResourceMetadata(required, existingCopy)
+		return existingCopy, false, nil
 	}
 
-	if klog.V(2).Enabled() {
-		klog.Infof("PrometheusRule %q changes: %v", namespace+"/"+required.GetName(), JSONPatchNoError(existing, toUpdate))
+	if klog.V(4).Enabled() {
+		klog.Infof("%s %q changes: %v", resourceGVR.String(), namespace+"/"+name, JSONPatchNoError(existing, existingCopy))
 	}
 
-	newObj, err := client.Resource(prometheusRuleGVR).Namespace(namespace).Update(ctx, toUpdate, metav1.UpdateOptions{})
-	if err != nil {
-		recorder.Warningf("PrometheusRuleUpdateFailed", "Failed to update PrometheusRule.monitoring.coreos.com/v1: %v", err)
-		return nil, true, err
-	}
-
-	recorder.Eventf("PrometheusRuleUpdated", "Updated PrometheusRule.monitoring.coreos.com/v1 because it changed")
-	return newObj, true, err
+	// Perform update if resource exists but different from the required (desired) one.
+	actual, err := client.Resource(resourceGVR).Namespace(namespace).Update(ctx, required, metav1.UpdateOptions{})
```

Comment:
Should this function not check if `required` has a `ResourceVersion` set and if so if it corresponds to the existing object before updating? Wouldn't this otherwise risk overwriting prior updates?

Reply:
I believe due to the fact that:

* the resource version cache updates [here](https://github.com/openshift/library-go/blob/e6bb5506390c06b624c7192bc41ed4fc4b517618/pkg/operator/resource/resourceapply/resource_cache.go#L149) are always sourced from the existing state and never the required one, and,
* an explicit required resource version that is not equal to the existing one will cause the Kube API to throw an error,

should ensure a valid and latest copy of the resource version is maintained in the cache, and used for skipping updates. WDYT?

Reply:
Right, the write happens [here](https://github.com/openshift/library-go/blob/e6bb5506390c06b624c7192bc41ed4fc4b517618/pkg/operator/resource/resourceapply/resource_cache.go#L117) where the cache key is dependent on the existing object's `ResourceVersion`. I've added a fail-fast check.

Reply:
Unfortunately, this doesn't seem to work because the server requires an RV to be set when performing an update. This manifests as (the operator is permanently broken):
```
{Failed to upgrade kube-apiserver, operator was degraded (KubeAPIServerStaticResources_SyncError): KubeAPIServerStaticResourcesDegraded: "assets/alerts/api-usage.yaml" (string): prometheusrules.monitoring.coreos.com "api-usage" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
KubeAPIServerStaticResourcesDegraded:   }
```

The issue is quite serious, as it will block upgrades to `4.18`. Please take a look at https://github.com/openshift/cluster-kube-apiserver-operator/pull/1750, especially the `e2e-aws-ovn-upgrade` job.

@rexagod , are you willing to work on a fix? I think we would like to have a test to prevent future bugs.
If the code is part of 4.17 I think we should back port the fix and the test.